### DISCUSSION TITLE
Using tar_combine to combine targets produced from tar_map

### DISCUSSION THREAD

**Discussion by jake-wittman** [ropensci/targets#987] [Help] [answered]
Hello! I am using static branching where each branch ultimately produces a vector. I would like to combine all of these branches into one data.frame, where each vector is a column in the new data.frame. I figured tar_combine() would be the way to do this, but I am having trouble understanding how to implement it. 

This is the code provided in the Targets tutorial document and part of my confusion stems from the fact that it is not organized how I am writing my _targets.R file:
```
# _targets.R file:
library(targets)
library(tarchetypes)
library(tibble)
random_seed <- tar_target(random_seed, seq_len(10))
mapped <- tar_map(
  unlist = FALSE, # Return a nested list from tar_map()
  values = tibble(method_function = rlang::syms(c("method1", "method2"))),
  tar_target(
    analysis,
    method_function("NIH", seed = random_seed),
    pattern = map(random_seed)
  ),
  tar_target(
    summary,
    summarize_analysis(analysis),
    pattern = map(analysis)
  )
)
combined <- tar_combine(
  combined_summaries,
  mapped[[2]],
  command = dplyr::bind_rows(!!!.x, .id = "method")
)
list(random_seed, mapped, combined)
```
I have my _targets.R file structured so each target is just an item in a list, which is how the walkthrough chapter in the tutorial shows how to structure the _targets.R file . So to rewrite the example from the static branching chapter, my _targets.R file would look like this:
```
library(targets)
library(tarchetypes)
library(tibble)
list(
  tar_target(random_seed, seq_len(10)),
  tar_map(
    unlist = FALSE, # Return a nested list from tar_map()
    values = tibble(method_function = rlang::syms(c("method1", "method2"))),
    tar_target(
      analysis,
      method_function("NIH", seed = random_seed),
      pattern = map(random_seed)
    ),
    tar_target(
      summary,
      summarize_analysis(analysis),
      pattern = map(analysis)
    )
  ),
  tar_combine(
    combined_summaries,
    mapped[[2]],
    command = dplyr::bind_rows(!!!.x, .id = "method")
  )
)
```
When I run tar_visnetwork() using the tutorial example, the visualization produces just fine. When I restructure code to use the format I'm used to working with tar_visnetwork() fails and says "object 'mapped' not found".  I suppose my question is: is there a way to use tar_combine() in the list formatted targets pipeline? Or do I need to rewrite my _targets.R file to be structured like the example.
Edit: fixed code formatting

**Comment by wlandau** [comment]
The error happens because the ellipsis  argument of tar_combine() should be target objects or lists of target objects. In your second example, you are still passing an object called ‘mapped’, but the _targets.R file does not define it.

To aggregate multiple vector targets into a data frame, you could set the command argument equal to data.frame(!!!.x) in tar_combine(). 

I suggest you have a look at https://docs.ropensci.org/tarchetypes/reference/tar_combine.html.

**Reply by jake-wittman** [reply]
Sorry I made an error when I posted my example - I realize that 'mapped' should be replaced, I just failed to do so in that example. If I replace 'mapped' with 'analysis' as in:

```
list(
  tar_target(random_seed, seq_len(10)),
  tar_map(
    unlist = FALSE, # Return a nested list from tar_map()
    values = tibble(method_function = rlang::syms(c("method1", "method2"))),
    tar_target(
      analysis,
      method_function("NIH", seed = random_seed),
      pattern = map(random_seed)
    ),
    tar_target(
      summary,
      summarize_analysis(analysis),
      pattern = map(analysis)
    )
  ),
  tar_combine(
    combined_summaries,
    analysis,
    command = dplyr::bind_rows(!!!.x, .id = "method")
  )
)
```
I get the error "object 'analysis' not found". If I understand what you wrote in your reply correctly, putting in a previous target (analysis) where the ellipsis argument is should work? I also tried replacing "analysis" with the actual full name of the two objects. Replacing "analysis" with "analysis_method1, analysis_method2" does not work either, and gives the same error. 

If I replace analysis with 'summary' I get the error: "Found a non-target object in the target list. The target script file (default: _targets.R) must end with a list of tar_target() objects (recommended) or a tar_pipeline() object (deprecated).

**Reply by wlandau** [reply]
The ellipsis argument expects one or more target objects defined in the R session, not the names of targets. A target object is the return value of tar_target(), and the target name is part of the target object. If you write ‘target_object <- tar_target(name = target_name, …)’, then tar_combine() can accept ‘target_object’, but not ‘target_name’. tar_combine() needs the actual full target object.

**Comment by jake-wittman** [comment, marked as answer]
So, if I understand correctly, there isn't a way to make use of tar_combine in the list of target objects without first explicitly defining the target object in R. The closest I can get is to give the targets object that is to be combined a name outside the list, then write the rest of my targets inside the list and add the target object to be combined to the list, like this:

```
# _targets.R file:
library(targets)
library(tarchetypes)
library(tibble)

mapped <- tar_map(
  unlist = FALSE, # Return a nested list from tar_map()
  values = tibble(method_function = rlang::syms(c("method1", "method2"))),
  tar_target(
    analysis,
    method_function("NIH", seed = random_seed),
    pattern = map(random_seed)
  ),
  tar_target(
    summary,
    summarize_analysis(analysis),
    pattern = map(analysis)
  )
)
list( 
 tar_target(random_seed, seq_len(10)),
 mapped,
 tar_combine(
  combined_summaries,
  mapped[[2]],
  command = dplyr::bind_rows(!!!.x, .id = "method")
))
```

I apologize if my language isn't the most precise or even correct in how I'm describing this. I guess at the root of my question is if there is a "style-guide" for targets. Should I be defining all of my target objects in the R session then calling those names in a list (like in the tar_combine() example from the static branching chapter)? Or is that only necessary to use something like `tar_combine()`, and I should otherwise just create my targets with `tar_target()` within the `list()` function?

**Reply by wlandau** [reply]
> So, if I understand correctly, there isn't a way to make use of tar_combine in the list of target objects without first explicitly defining the target object in R. The closest I can get is to give the targets object that is to be combined a name outside the list, then write the rest of my targets inside the list and add the target object to be combined to the list, like this:

Yes, that is how it is intended to be used. `tar_combine()` accepts target objects because they are available and convenient in the most typical case, when the targets you want to combine are generated with `tar_map()` (or `tar_eval()`).

> Should I be defining all of my target objects in the R session then calling those names in a list (like in the tar_combine() example from the static branching chapter)? Or is that only necessary to use something like tar_combine(), and I should otherwise just create my targets with tar_target() within the list() function?

I recommend generating pieces of the target list separately when there is a need to use those target objects later in `_targets.R`, as with `tar_combine()`. Otherwise, it is up to you how to write the target list. It is good practice to make it clear and concise using custom functions like `summarize_analysis()`, which make the individual target commands clear and concise. If the pipeline is not too big, you can write all the targets in the `_targets.R` file itself. Or if the number of targets gets extremely large, you may choose to write pieces of the target list in separate R script, then combine them in `_targets.R`. Whatever is clearest for the use case.

**Comment by zdravis** [comment]
@wlandau in the above example, can you provide any help with using `dplyr::left_join()` instead of `dplyr::bind_rows()`? Specifically, I get the error "problematic arguments: target1_name = target1_name and target2_name = target2_name". I am wondering if the `!!!.x` isn't working with left_join and if there is a way to join rows and build a wider dataset as opposed to unioned long dataset. thanks!

**Reply by wlandau** [reply]
For `left_join()`, which only takes two arguments `x` and `y`, I suggest doing the join manually instead of with `tar_combine()`. `tar_combine()` is meant for functions like `list()` and `bind_rows()` which accept any number of arbitrarily named arguments through the ellipsis `...`.

**Reply by zdravis** [reply]
great, thank you for the speedy reply!

**Comment by karimn** [comment]
I came upon this issue when trying to fix a tar_map/tar_combine issue I was seeing. I fixed it now, but I wanted to give a heads up. If you use `purrr::lst` instead of the base R `list` for your targets, the `.x` arg in tar_combine results in an error.
